### PR TITLE
feat(runtime): ChatExecutor with cooldown-based provider fallback

### LINKED ISSUE #1063: Phase 1.11: Agent loop integration (ChatExecutor + model fallback)
## Summary

Connect the Gateway message flow to the existing LLM + tool execution pipeline. This is the central integration point where inbound messages flow through identity resolution, session management, system prompt assembly, LLM execution with tool calls, and outbound response delivery. The key new component is `ChatExecutor` — a message-oriented variant of the existing `LLMTaskExecutor` that takes `GatewayMessage` + conversation history and produces string responses (not `bigint[]`). ChatExecutor supports model fallback chains (if primary provider fails, try next), session-level token budget tracking, streaming output, and dynamic system prompt assembly from workspace files + skills + memory context.

**Target**: Runtime Package
**Complexity**: HIGH
**Estimated files**:
- `runtime/src/llm/chat-executor.ts` (new)
- `runtime/src/llm/chat-executor.test.ts` (new)
- `runtime/src/llm/index.ts` (modify)
- `runtime/src/gateway/gateway.ts` (modify — wire message flow)
- `runtime/src/index.ts` (modify)

## Dependencies

Blocked by:
- #1053 - Phase 1.1: Gateway core process (requires Gateway class)
- #1051 - Phase 1.2: Unified message format (requires GatewayMessage type)
- #1055 - Phase 1.6: Session management (requires Session manager)
- #1056 - Phase 1.7: Lifecycle hook system (requires hook dispatch)
- #1052 - Phase 1.8: Slash commands handler (requires slash command context)
- #1057 - Phase 1.9: Cross-channel identity linking (requires identity resolver)

### MUST Requirements
- [ ] `ChatExecutor` class with `execute()` method
  - Implementation: Takes `ChatExecuteParams` (message, history, systemPrompt, sessionId, capabilities), returns `ChatExecutorResult`
- [ ] Model fallback chain with cooldown-based recovery
  - Implementation: Try primary provider, on error try fallbacks in order; failed provider enters cooldown, skipped until expiry
- [ ] Tool call loop integration
  - Implementation: Reuse existing `LLMTaskExecutor` tool call loop logic; dispatch `tool:before` and `tool:after` hooks per call
- [ ] Session-level token budget tracking
  - Implementation: Track cumulative tokens per sessionId; reject execution if budget exceeded; provide `getSessionTokenUsage()` and `resetSessionTokens()`
- [ ] System prompt assembly with optional skill/memory injection
  - Implementation: Assemble in order: AGENT.md → SOUL.md → USER.md → TOOLS.md → injected skills (if SkillInjector provided) → relevant memories (if MemoryRetriever provided)
- [ ] Streaming support via optional callback
  - Implementation: If `onStreamChunk` provided, use provider's streaming API to dispatch chunks
- [ ] `allowedTools` allowlist for defense-in-depth
  - Implementation: Filter tool calls to only allowed tools; reject calls to non-listed tools

### SHOULD Requirements
- [ ] Provider cooldown management
  - Implementation: Track failed providers with cooldown expiry; auto-retry after cooldown
- [ ] `ToolCallRecord` capturing all executions
  - Implementation: Record tool name, args, result, isError, durationMs for each call; include in `ChatExecutorResult`
- [ ] Integration with existing `LLMProvider` adapters
  - Implementation: Reuse Grok, Anthropic, Ollama adapters; no new provider implementations needed
- [ ] Comprehensive test coverage
  - Implementation: 18+ test cases covering happy path, fallback, tool calls, token budget, streaming, skill/memory injection

### MAY Requirements
- [ ] Hook dispatch for tool execution observability (optional if Phase 1.7 not blocked)
  - Implementation: Call hook handlers for `tool:before` and `tool:after` around each tool call
- [ ] Session-level model stickiness (optional improvement)
  - Implementation: Remember which fallback succeeded for a session, prefer it on next execution for cache warmth

### ChatExecutor Class Structure

```typescript
// runtime/src/llm/chat-executor.ts

export interface ChatExecutorConfig {
  readonly provider: LLMProvider;
  readonly fallbacks?: readonly LLMProvider[];
  readonly toolHandler?: ToolHandler;
  readonly maxToolRounds?: number;
  readonly onStreamChunk?: (chunk: string) => void;
  readonly skillInjector?: SkillInjector;
  readonly memoryRetriever?: MemoryRetriever;
  readonly allowedTools?: readonly string[];
  readonly sessionTokenBudget?: number;
  readonly providerCooldownMs?: number;
}

export interface ChatExecutorResult {
  readonly content: string;
  readonly provider: string;
  readonly usedFallback: boolean;
  readonly toolCalls: readonly ToolCallRecord[];
  readonly tokenUsage: { prompt: number; completion: number; total: number };
  readonly durationMs: number;
}

export interface ToolCallRecord {
  readonly toolName: string;
  readonly args: Record<string, unknown>;
  readonly result: string;
  readonly isError: boolean;
  readonly durationMs: number;
}

export interface SkillInjector {
  inject(message: string, history: readonly LLMMessage[], capabilities: bigint): Promise<string>;
}

export interface MemoryRetriever {
  retrieve(message: string, sessionId: string, limit?: number): Promise<string>;
}

export interface ChatExecuteParams {
  readonly message: GatewayMessage;
  readonly history: readonly LLMMessage[];
  readonly systemPrompt: string;
  readonly sessionId: string;
  readonly capabilities?: bigint;
}

export class ChatExecutor {
  constructor(config: ChatExecutorConfig);
  execute(params: ChatExecuteParams): Promise<ChatExecutorResult>;
  getSessionTokenUsage(sessionId: string): number;
  resetSessionTokens(sessionId: string): void;
}
```

### Key Implementation Patterns

**Reuse from existing code:**
- `LLMProvider` interface — Grok, Anthropic, Ollama all implement this
- `LLMTaskExecutor` tool call loop logic (5 call rounds with `toolHandler.execute()`)
- `ToolRegistry` and `ToolHandler` from Phase 1.4
- `LLMMessage` type from Phase 1.2
- `GatewayMessage` type from Phase 1.2

**Model fallback flow:**
```typescript
let lastError: Error | null = null;
const providers = [config.provider, ...(config.fallbacks || [])];

for (const provider of providers) {
  if (this.isInCooldown(provider.name)) continue;
  
  try {
    const result = await this.executeWithProvider(provider, params);
    return result;
  } catch (error) {
    lastError = error;
    this.startCooldown(provider.name, config.providerCooldownMs ?? 60000);
  }
}

throw lastError ?? new Error('All providers failed');
```

**Token tracking:**
```typescript
private tokenUsageBySession = new Map<string, number>();

getSessionTokenUsage(sessionId: string): number {
  return this.tokenUsageBySession.get(sessionId) ?? 0;
}

// After each execution:
const currentUsage = this.getSessionTokenUsage(params.sessionId);
const newUsage = currentUsage + result.tokenUsage.total;
if (config.sessionTokenBudget && newUsage > config.sessionTokenBudget) {
  throw new Error(`Token budget exceeded: ${newUsage} > ${config.sessionTokenBudget}`);
}
this.tokenUsageBySession.set(params.sessionId, newUsage);
```

### Gateway Integration (sketch)

In `runtime/src/gateway/gateway.ts`:
```typescript
async onMessage(message: GatewayMessage): Promise<void> {
  // ... resolve identity, route to workspace, get session ...
  
  const history = await this.session.getHistory();
  const systemPrompt = this.assembleSystemPrompt(workspace);
  
  const result = await this.chatExecutor.execute({
    message,
    history,
    systemPrompt,
    sessionId: session.id,
    capabilities: this.agentCapabilities,
  });
  
  const response: OutboundMessage = {
    channel: message.channel,
    targetId: message.senderId,
    content: result.content,
  };
  
  await this.dispatchHook('message:outbound', response);
  await this.channels[message.channel].send(response);
  await this.session.appendMessage(response);
}
```

## Test Plan

- [ ] Happy path: execute with primary provider returns response with provider name
- [ ] Model fallback: primary provider fails (throw error), secondary provider succeeds
- [ ] `usedFallback: true` when fallback used, `false` when primary succeeds
- [ ] Provider cooldown: failed provider skipped for cooldown duration
- [ ] Provider cooldown expires: provider retried after cooldown elapsed
- [ ] Tool call loop: single tool call round executes correctly
- [ ] Tool call loop: multiple rounds chain correctly with context
- [ ] `maxToolRounds` enforced: execution stops after max rounds even if tool requests continuation
- [ ] `allowedTools` filtering: non-allowed tool calls rejected
- [ ] Token budget: execution rejected when cumulative tokens exceed budget
- [ ] `getSessionTokenUsage()` returns cumulative tokens across multiple executions
- [ ] `resetSessionTokens()` clears token tracking for a session
- [ ] Streaming callback receives chunks if `onStreamChunk` provided
- [ ] `SkillInjector.inject()` content appended to system prompt
- [ ] `MemoryRetriever.retrieve()` content appended to system prompt
- [ ] `ChatExecutorResult.toolCalls` includes all tool call records with args, results, durations
- [ ] Empty history works (first message in session)
- [ ] All providers fail: throws error with details from each attempt
- [ ] Concurrent executions in different sessions don't interfere with token tracking

## Similar Existing Code

Reference these files for patterns:
- `runtime/src/llm/executor.ts` — `LLMTaskExecutor` (tool call loop, provider execution pattern)
- `runtime/src/llm/types.ts` — `LLMMessage`, `LLMProvider` interfaces
- `runtime/src/tools/registry.ts` — `ToolRegistry`, `ToolHandler` interface
- `runtime/src/agent/manager.ts` — Session-based state management pattern
- `runtime/src/memory/*/backend.ts` — Session ID tracking pattern

## Notes

- `SkillInjector` and `MemoryRetriever` are optional interfaces passed to config — implement as `undefined` pass-through until Phase 3.3/5.5
- Do NOT implement workspace file loading (AGENT.md, etc.) — Gateway assembles and passes pre-assembled `systemPrompt` to ChatExecutor
- Do NOT implement approval gates or dangerous action checks — Phase 1.10 handles that
- ChatExecutor is stateless per execution; state (token budget, cooldowns) managed internally
- Reuse existing provider implementations — no new adapters needed
- Hook dispatch (`tool:before`/`tool:after`) only if Phase 1.7 blocking is resolved

---

**Document Reference**: [AgenC Roadmap](https://github.com/tetsuo-ai/AgenC/blob/main/docs/ROADMAP.md)
**Section**: 1.11 - Agent Loop Integration

### PR DESCRIPTION
**Pull request by 7etsuo**
## Summary

- Add `ChatExecutor` — a message-oriented LLM executor for the gateway pipeline with cooldown-based provider fallback and session-level token budget tracking
- Implements tool call loop with allowlist enforcement, JSON validation, and `ToolCallRecord` tracking
- Includes optional `SkillInjector` and `MemoryRetriever` injection points for future phases
- Backfills missing top-level barrel exports (`FallbackLLMProvider`, `LLMAuthenticationError`, `LLMServerError`, `mapLLMError`, `validateToolCall`)

## Test plan

- [x] 26 new test cases in `chat-executor.test.ts` covering basic operation, fallback, cooldown (with fake timers), tool loop, token budget, injection, and edge cases
- [x] All 112 LLM tests pass (8 files, zero regressions)
- [x] TypeScript type-check passes (no new errors)

Closes #1063

### CODE REVIEW

File: runtime/src/llm/chat-executor.ts [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+ */
+export class ChatBudgetExceededError extends RuntimeError {
+  public readonly sessionId: string;
+  public readonly used: number;
+  public readonly limit: number;
+
+  constructor(sessionId: string, used: number, limit: number) {
+    super(
+      `Token budget exceeded for session "${sessionId}": ${used}/${limit}`,
+      RuntimeErrorCodes.LLM_PROVIDER_ERROR,
```
Comment (7etsuo) [inline runtime/src/llm/chat-executor.ts]: **High: `ChatBudgetExceededError` reuses `LLM_PROVIDER_ERROR` code.**

This error is a budget/policy error, not a provider error. Using the same code makes it impossible to programmatically distinguish budget exhaustion from actual provider failures via the `code` field.

Suggested fix: Add a new `RuntimeErrorCode` like `CHAT_BUDGET_EXCEEDED`, or at minimum use a different existing code (e.g., `RATE_LIMIT_ERROR` is semantically closer).

File: runtime/src/llm/chat-executor.ts [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+    // Append history and user message
+    messages.push(...history);
+    messages.push({ role: 'user', content: message.content });
+
+    // First LLM call
+    const cumulativeUsage: LLMUsage = { promptTokens: 0, completionTokens: 0, totalTokens: 0 };
+    const allToolCalls: ToolCallRecord[] = [];
+
+    let { response, providerName, usedFallback } = await this.callWithFallback(messages);
```
Comment (7etsuo) [inline runtime/src/llm/chat-executor.ts]: **High: `SkillInjector` and `MemoryRetriever` types not exported from barrels.**

Consumers building `ChatExecutorConfig` with a `skillInjector` or `memoryRetriever` can't reference these types when importing from `@agenc/runtime`. Either export them as type-only from the barrel, or use inline function signatures in the config interface.

The comment says "not exported until Phase 3.3/5.5" — but the config accepts them NOW, so consumers need the types NOW.

File: runtime/src/llm/chat-executor.ts [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        }
+
+        // Success — clear cooldown
+        this.cooldowns.delete(provider.name);
+
+        return {
+          response,
+          providerName: provider.name,
+          usedFallback: i > 0,
+        };
```
Comment (7etsuo) [inline runtime/src/llm/chat-executor.ts]: **Medium: Missing test — all providers in cooldown simultaneously.**

If all providers are in cooldown, `callWithFallback` skips everyone. `lastError` would be `undefined` (no provider tried, no error caught), so it throws a generic "All providers failed" with no context about cooldowns. Should test this and throw a more descriptive error like "All providers in cooldown".

File: runtime/src/llm/chat-executor.ts [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    this.toolHandler = config.toolHandler;
+    this.maxToolRounds = config.maxToolRounds ?? 10;
+    this.onStreamChunk = config.onStreamChunk;
+    this.allowedTools = config.allowedTools ? new Set(config.allowedTools) : null;
+    this.sessionTokenBudget = config.sessionTokenBudget;
+    this.cooldownMs = config.providerCooldownMs ?? 60_000;
+    this.maxCooldownMs = config.maxCooldownMs ?? 300_000;
+    this.maxTrackedSessions = config.maxTrackedSessions ?? 10_000;
+    this.skillInjector = config.skillInjector;
+    this.memoryRetriever = config.memoryRetriever;
```
Comment (7etsuo) [inline runtime/src/llm/chat-executor.ts]: **Medium: Constructor does not validate `providerCooldownMs` or `maxCooldownMs`.**

Passing `providerCooldownMs: -1` or `maxCooldownMs: 0` would cause unexpected cooldown behavior (negative cooldowns, `Math.min` always returns 0). Add validation or clamp to reasonable minimums.

File: runtime/src/index.ts [hunk 5]
```diff
@@ -465,16 +465,31 @@ export {
...
+  mapLLMError,
   // Response converter
   responseToOutput,
+  // Utilities
+  validateToolCall,
   // LLM Task Executor
   LLMTaskExecutor,
   type LLMTaskExecutorConfig,
+  // Fallback provider
+  FallbackLLMProvider,
```
Comment (7etsuo) [inline runtime/src/index.ts]: **Medium: Backfilling exports for `FallbackLLMProvider`, `LLMAuthenticationError`, `LLMServerError`, `mapLLMError`, `validateToolCall` is unrelated to ChatExecutor.**

These existed in `llm/index.ts` but were missing from the top-level barrel. Adding them in THIS PR is scope creep — they should be a separate commit or at least called out in the PR description.
